resources/tentative/complex-static-html/src/big-dom-generator/gen-css.js

```diff
@@ -1,238 +1,197 @@
+    const randomNumber = random.float();
+    let accumulatedProbability = 0;
+    for (let i = 0; i < probabilities.length; i++) {
+        accumulatedProbability += probabilities[i];
         // prettier-ignore
-        if (randNum <= accumProb)
+        if (randomNumber <= accumulatedProbability)
             return options[i];
     }
     return options[options.length - 1];
 };
 
-const buildMatchingSelector = (depth, index, oldCombinator, selLen, maxLen) => {
-    // prettier-ignore
-    if (selLen >= maxLen)
-        return "";
+const buildSelectors = (element, depth, oldCombinator, selectorLength, maxSelectorLength, isMatching) => {
+    // if nonMatching we add a view-<random-index> class selector that is guaranteed to not have targeted children in the todoMVC
+    if ((!isMatching && !depth) || selectorLength >= maxSelectorLength || !element)
+        return isMatching ? "" : `.view-${random.randRange(0, NUM_TODOS_TO_INSERT_IN_HTML)}${oldCombinator}`;
 
-    const getSelector = randomWeighted([getClassname, getType, () => "*"], [0.6, 0.3, 0.1]);
-    const selector = getSelector(depth, index);
-    const combinator = chooseCombinator(depth, index);
+    const getSelector = randomWeighted([getClassname, getElementType, () => "*"], [0.6, 0.3, 0.1]);
+    const selector = getSelector(element);
 
-    // prettier-ignore
-    if (!depth)
+    if (isMatching && !depth)
         return `${selector}${oldCombinator}`;
 
+    const combinator = chooseCombinator(element);
     const nextDepth = getNextDepth(combinator, depth);
-    const nextIndex = getNextIndex(combinator, nextDepth, index);
-    return buildMatchingSelector(nextDepth, nextIndex, combinator, selLen + 1, maxLen) + selector + oldCombinator;
+    const nextElement = getRandomElement(combinator, element, depth, nextDepth);
+
+    return buildSelectors(nextElement, nextDepth, combinator, selectorLength + 1, maxSelectorLength, isMatching) + selector + oldCombinator;
 };
 
-// Returns a non-matching selector for the element.
-// We kept the selector from matching by adding the `.just-span` class to the left-most selector or
-// by adding a classname from one of its children.
-// TODO: Is there a better way to ensure the selector is non-matching?
-const buildNonMatchingSelector = (depth, index, oldCombinator, selLen, badSelector) => {
-    // If we are in the top node, we are done.
-    // prettier-ignore
-    if (!depth)
-        return `.just-span${ oldCombinator}`;
-
-    // If we've reached the target length, pick a random classname from its children.
-    const getSelector = randomWeighted([getClassname, getType, () => "*"], [0.6, 0.3, 0.1]);
-    const selector = getSelector(depth, index);
-    if (selLen === badSelector) {
-        const wrongDepth = random.randRange(Math.min(depth + 1, 7), 8);
-        const wrongSelector = getClassname(wrongDepth, 0);
-        return selector + wrongSelector + oldCombinator;
-    }
+const ANGULAR_VIEW_DEPTH = 8;
```

Comment:
Since we now have JSDOM running in the script, is it feasible / worthwhile to derive these numbers from the HTML instead of having constants (i.e. querySelector for the relevant element and count the number of parents, or something). It looks like the only consumer is called from a function that has access to the element (`getInitialDepth`) so perhaps these numbers could be derived directly from the element (and it may not even need to know if we're running in Angular or not)

Reply:
Agreed.